spring-data-rest deserialization support

Testing the new spring-data-rest deserialization i created a PoC

https://github.com/EugenMayer/java-spring-pocs/blob/master/src/main/java/de/kontextwork/poc/spring/blaze/fullapp/subject/group/controller/GroupController.java

With the test here

https://github.com/EugenMayer/java-spring-pocs/blob/master/src/test/java/de/kontextwork/poc/spring/blaze/fullapp/subject/group/controller/GroupControllerTest.java#L107

I am running against the current master snapshot for now.

Steps i took to integrate:

- add "com.blazebit:blaze-persistence-integration-spring-data-rest:${blazePersistence}"

When trying to run the tests i get an exception about the sorter 

```
Caused by: org.springframework.beans.factory.support.BeanDefinitionOverrideException: Invalid bean definition with name 'sortResolver' defined in class path resource [org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.class]: Cannot register bean definition [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=org.springframework.data.rest.webmvc.config.RepositoryRestMvcConfiguration; factoryMethodName=sortResolver; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.class]] for bean 'sortResolver': There is already [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=com.blazebit.persistence.spring.data.rest.impl.BlazePersistenceWebConfiguration; factoryMethodName=sortResolver; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [com/blazebit/persistence/spring/data/rest/impl/BlazePersistenceWebConfiguration.class]] bound.
```

I guess this comes from `BlazePersistenceWebConfiguration` being auto exposed and auto configured.
Considering that i do not define my own sorter - i do actually not override anything here:
https://github.com/EugenMayer/java-spring-pocs/blob/master/src/main/java/de/kontextwork/poc/spring/configuration/RestConfiguration.java#L23

I assume this exception would be thrown on every single application - not just this poc. I think we are trying to auto-reconfigure the spring default which is defined in `SpringDataWebConfiguration`

and is actually already the same. In this case either we are doing the integration the wrong way or in this specific case i would tend to go with you should not define this bean at all since it is the default and the spring one also adds `customizeSortResolver(sortResolver);` which is probably a different hook to customize the sorter in general - in any way we are defining a `@Bean` which is already the default -> IMHO remove it

--- a/integration/spring-data/webmvc/src/main/java/com/blazebit/persistence/spring/data/webmvc/KeysetPageableArgumentResolver.java
+++ b/integration/spring-data/webmvc/src/main/java/com/blazebit/persistence/spring/data/webmvc/KeysetPageableArgumentResolver.java
@@ -1,0 +1,143 @@
+/*
+ * Copyright 2014 - 2020 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.spring.data.webmvc;
+
+import com.blazebit.persistence.spring.data.repository.KeysetPageable;
+import com.blazebit.persistence.spring.data.webmvc.impl.KeysetPageableHandlerMethodArgumentResolver;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.web.PageableArgumentResolver;
+import org.springframework.data.web.PageableDefault;
+import org.springframework.web.context.request.NativeWebRequest;
+
+/**
+ * Argument resolver to extract a {@link KeysetPageable} object from a {@link NativeWebRequest} for a particular
+ * {@link org.springframework.core.MethodParameter}. {@link KeysetPageable} resolution yields either
+ * in a {@link Pageable} object or {@literal null} if {@link KeysetPageable} cannot be resolved.
+ *
+ * @author Moritz Becker
+ * @since 1.4.1
+ */
+public interface KeysetPageableArgumentResolver extends PageableArgumentResolver {
+    /**
+     * Configures the {@link Pageable} to be used as fallback in case no {@link PageableDefault} or
+     * {@link PageableDefault} (the latter only supported in legacy mode) can be found at the method parameter to be
+     * resolved.
+     * <p>
+     * If you set this to {@literal Optional#empty()}, be aware that you controller methods will get {@literal null}
+     * handed into them in case no {@link Pageable} data can be found in the request. Note, that doing so will require you
+     * supply bot the page <em>and</em> the size parameter with the requests as there will be no default for any of the
+     * parameters available.
+     *
+     * @param fallbackPageable the {@link Pageable} to be used as general fallback.
+     */
+    void setFallbackPageable(Pageable fallbackPageable);
+
+    /**
+     * @param fallbackPageable the {@link KeysetPageable} to be used as general fallback.
+     * @see KeysetPageableArgumentResolver#setFallbackPageable(Pageable)
+     */
+    void setFallbackPageable(KeysetPageable fallbackPageable);
+
+    /**
+     * Configures the maximum page size to be accepted. This allows to put an upper boundary of the page size to prevent
+     * potential attacks trying to issue an {@link OutOfMemoryError}. Defaults to {@link org.springframework.data.web.PageableHandlerMethodArgumentResolver#DEFAULT_MAX_PAGE_SIZE}.
+     *
+     * @param maxPageSize the maxPageSize to set
+     */
+    void setMaxPageSize(int maxPageSize);
+
+    /**
+     * Configures the parameter name to be used to find the page number in the request. Defaults to {@code page}.
+     *
+     * @param pageParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setPageParameterName(String pageParameterName);
+
+    /**
+     * Configures the parameter name to be used to find the page size in the request. Defaults to {@code size}.
+     *
+     * @param sizeParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setSizeParameterName(String sizeParameterName);
+
+    /**
+     * Configures a general prefix to be prepended to the page number and page size parameters. Useful to namespace the
+     * property names used in case they are clashing with ones used by your application. By default, no prefix is used.
+     *
+     * @param prefix the prefix to be used or {@literal null} to reset to the default.
+     */
+    void setPrefix(String prefix);
+
+    /**
+     * The delimiter to be used between the qualifier and the actual page number and size properties. Defaults to
+     * {@code _}. So a qualifier of {@code foo} will result in a page number parameter of {@code foo_page}.
+     *
+     * @param qualifierDelimiter the delimiter to be used or {@literal null} to reset to the default.
+     */
+    void setQualifierDelimiter(String qualifierDelimiter);
+
+    /**
+     * Configures whether to expose and assume 1-based page number indexes in the request parameters. Defaults to
+     * {@literal false}, meaning a page number of 0 in the request equals the first page. If this is set to
+     * {@literal true}, a page number of 1 in the request will be considered the first page.
+     *
+     * @param oneIndexedParameters the oneIndexedParameters to set
+     */
+    void setOneIndexedParameters(boolean oneIndexedParameters);
+
+    /**
+     * Configures the parameter name to be used to find the offset in the request. Defaults to {@link KeysetPageableHandlerMethodArgumentResolver#DEFAULT_OFFSET_PARAMETER}.
+     *
+     * @param offsetParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setOffsetParameterName(String offsetParameterName);
+
+    /**
+     * Configures the parameter name to be used to find the previous offset in the request. Defaults to {@link KeysetPageableHandlerMethodArgumentResolver#DEFAULT_PREVIOUS_OFFSET_PARAMETER}.
+     *
+     * @param previousOffsetParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setPreviousOffsetParameterName(String previousOffsetParameterName);
+
+    /**
+     * Configures the parameter name to be used to find the previous page in the request. Defaults to {@link KeysetPageableHandlerMethodArgumentResolver#DEFAULT_PREVIOUS_PAGE_PARAMETER}.
+     *
+     * @param previousPageParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setPreviousPageParameterName(String previousPageParameterName);
+
+    /**
+     * Configures the parameter name to be used to find the previous size in the request. Defaults to {@link KeysetPageableHandlerMethodArgumentResolver#DEFAULT_PREVIOUS_SIZE_PARAMETER}.
+     *
+     * @param previousSizeParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setPreviousSizeParameterName(String previousSizeParameterName);
+
+    /**
+     * Configures the parameter name to be used to find the lowest keyset in the request. Defaults to {@link KeysetPageableHandlerMethodArgumentResolver#DEFAULT_LOWEST_PARAMETER}.
+     *
+     * @param lowestParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setLowestParameterName(String lowestParameterName);
+
+    /**
+     * Configures the parameter name to be used to find the highest keyset in the request. Defaults to {@link KeysetPageableHandlerMethodArgumentResolver#DEFAULT_HIGHEST_PARAMETER}.
+     *
+     * @param highestParameterName the parameter name to be used, must not be {@literal null} or empty.
+     */
+    void setHighestParameterName(String highestParameterName);
+}

--- a/integration/spring-data/webmvc/src/main/java/com/blazebit/persistence/spring/data/webmvc/impl/BlazePersistenceWebConfiguration.java
+++ b/integration/spring-data/webmvc/src/main/java/com/blazebit/persistence/spring/data/webmvc/impl/BlazePersistenceWebConfiguration.java
@@ -16,6 +16,7 @@
 
 package com.blazebit.persistence.spring.data.webmvc.impl;
 
+import com.blazebit.persistence.spring.data.webmvc.KeysetPageableArgumentResolver;
 import com.blazebit.persistence.spring.data.webmvc.impl.json.EntityViewAwareMappingJackson2HttpMessageConverter;
 import com.blazebit.persistence.view.EntityViewManager;
 import org.springframework.beans.factory.ObjectFactory;
@@ -54,13 +55,14 @@ public class BlazePersistenceWebConfiguration extends WebMvcConfigurerAdapter {
     }
 
     @Bean
-    public KeysetPageableHandlerMethodArgumentResolver keysetPageableResolver() {
+    public KeysetPageableArgumentResolver keysetPageableResolver() {
         return new KeysetPageableHandlerMethodArgumentResolver(sortResolver, conversionService.getObject());
     }
 
     @Override
     public void addArgumentResolvers(List<HandlerMethodArgumentResolver> argumentResolvers) {
-        argumentResolvers.add(keysetPageableResolver());
+        // Add it to the beginning so it has precedence over the builtin
+        argumentResolvers.add(0, keysetPageableResolver());
     }
 
     @Override

--- a/integration/spring-data/webmvc/src/main/java/com/blazebit/persistence/spring/data/webmvc/impl/KeysetPageableHandlerMethodArgumentResolver.java
+++ b/integration/spring-data/webmvc/src/main/java/com/blazebit/persistence/spring/data/webmvc/impl/KeysetPageableHandlerMethodArgumentResolver.java
@@ -22,6 +22,7 @@ import com.blazebit.persistence.KeysetPage;
 import com.blazebit.persistence.spring.data.repository.KeysetPageRequest;
 import com.blazebit.persistence.spring.data.repository.KeysetPageable;
 import com.blazebit.persistence.spring.data.webmvc.KeysetConfig;
+import com.blazebit.persistence.spring.data.webmvc.KeysetPageableArgumentResolver;
 import com.blazebit.reflection.ReflectionUtils;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
@@ -60,7 +61,7 @@ import java.util.concurrent.ConcurrentMap;
  * @author Christian Beikov
  * @since 1.2.0
  */
-public class KeysetPageableHandlerMethodArgumentResolver extends PageableHandlerMethodArgumentResolver {
+public class KeysetPageableHandlerMethodArgumentResolver extends PageableHandlerMethodArgumentResolver implements KeysetPageableArgumentResolver {
 
     private static final String DEFAULT_OFFSET_PARAMETER = "offset";
     private static final String DEFAULT_PREVIOUS_OFFSET_PARAMETER = "prevOffset";
@@ -157,6 +158,7 @@ public class KeysetPageableHandlerMethodArgumentResolver extends PageableHandler
         setFallbackPageable((KeysetPageable) fallbackPageable);
     }
 
+    @Override
     public void setFallbackPageable(KeysetPageable fallbackPageable) {
         this.fallbackPageable = fallbackPageable;
     }
@@ -166,50 +168,56 @@ public class KeysetPageableHandlerMethodArgumentResolver extends PageableHandler
         return fallbackPageable == null ? false : fallbackPageable.equals(pageable);
     }
 
-    public String getOffsetParameterName() {
+    protected String getOffsetParameterName() {
         return offsetParameterName;
     }
 
+    @Override
     public void setOffsetParameterName(String offsetParameterName) {
         this.offsetParameterName = offsetParameterName;
     }
 
-    public String getPreviousOffsetParameterName() {
+    protected String getPreviousOffsetParameterName() {
         return previousOffsetParameterName;
     }
 
+    @Override
     public void setPreviousOffsetParameterName(String previousOffsetParameterName) {
         this.previousOffsetParameterName = previousOffsetParameterName;
     }
 
-    public String getPreviousPageParameterName() {
+    protected String getPreviousPageParameterName() {
         return previousPageParameterName;
     }
 
+    @Override
     public void setPreviousPageParameterName(String previousPageParameterName) {
         this.previousPageParameterName = previousPageParameterName;
     }
 
-    public String getPreviousSizeParameterName() {
+    protected String getPreviousSizeParameterName() {
         return previousSizeParameterName;
     }
 
+    @Override
     public void setPreviousSizeParameterName(String previousSizeParameterName) {
         this.previousSizeParameterName = previousSizeParameterName;
     }
 
-    public String getLowestParameterName() {
+    protected String getLowestParameterName() {
         return lowestParameterName;
     }
 
+    @Override
     public void setLowestParameterName(String lowestParameterName) {
         this.lowestParameterName = lowestParameterName;
     }
 
-    public String getHighestParameterName() {
+    protected String getHighestParameterName() {
         return highestParameterName;
     }
 
+    @Override
     public void setHighestParameterName(String highestParameterName) {
         this.highestParameterName = highestParameterName;
     }

